File: LeetCode/00086. Partition List/00086.c
struct ListNode * partition(struct ListNode *head, int x) {
    struct ListNode *walker = head, *newWalker = NULL;
    struct ListNode *lessHead = NULL, *lessEnd = NULL;
    struct ListNode *greaterHead = NULL, *greaterEnd = NULL;
    
    while (walker != NULL) {
        newWalker = walker->next;
        walker->next = NULL;
        
        if (walker->val < x) {
            if (lessHead == NULL) {
                lessHead = walker;
                lessEnd  = walker;
            } else {
                lessEnd->next = walker;
                lessEnd = walker;
            }
        } else {
            if (greaterHead == NULL) {
                greaterHead = walker;
                greaterEnd  = walker;
            } else {
                greaterEnd->next = walker;
                greaterEnd = walker;
            }
        }
        
        walker = newWalker;
    }
    if (lessEnd != NULL) {
        lessEnd->next = greaterHead;
    }
    
    return (lessHead != NULL) ? lessHead : greaterHead;
}

File: LeetCode/00086. Partition List/00086.go
// https://leetcode.com/problems/partition-list

func partition(head *ListNode, x int) *ListNode {
    dummyA, dummyB := &ListNode{}, &ListNode{}
    a, b := dummyA, dummyB

    for head != nil {
        node := head
        head = head.Next
        node.Next = nil

        if node.Val < x {
            a.Next = node
            a = node
        } else {
            b.Next = node
            b = node
        }
    }
    a.Next = dummyB.Next

    return dummyA.Next
}
